File: approval-inbox/typescript/src/typesConfig.ts
/**
 * Loads approval_types.yaml. Re-loads on every call.
 */

import { existsSync, readFileSync } from "node:fs";
import { parse as parseYaml } from "yaml";

const FILE = "approval_types.yaml";

export interface ActionBlock {
  forward_to?: string;
  webhook?: string;
  reply_to_sender?: string;
}

export interface TypeConfig {
  type: string;
  description: string;
  classifier_hints: { senders?: string[]; keywords?: string[] };
  extract_fields: string[];
  approve: ActionBlock;
  decline: ActionBlock;
}

export function load(): TypeConfig[] {
  if (!existsSync(FILE)) return [];
  let data: any;
  try {
    data = parseYaml(readFileSync(FILE, "utf-8"));
  } catch (e: any) {
    console.warn(`  ! approval_types.yaml malformed: ${e.message}`);
    return [];
  }
  const out: TypeConfig[] = [];
  for (const entry of (data?.types ?? []) as any[]) {
    out.push({
      type: entry.type,
      description: entry.description ?? "",
      classifier_hints: entry.classifier_hints ?? {},
      extract_fields: entry.extract_fields ?? [],
      approve: entry.approve ?? {},
      decline: entry.decline ?? {},
    });
  }
  return out;
}

export function find(types: TypeConfig[], typeName: string): TypeConfig | null {
  return types.find(t => t.type === typeName) ?? null;
}

export function renderForPrompt(types: TypeConfig[]): string {
  if (!types.length) {
    return "(no request types configured — every email will be discarded)";
  }
  return types.map(t => {
    const senders = (t.classifier_hints.senders || []).join(", ") || "(no sender hints)";
    const keywords = (t.classifier_hints.keywords || []).join(", ") || "(no keyword hints)";
    const fields = t.extract_fields.join(", ") || "(no fields)";
    return (
      `## ${t.type}\n` +
      `  - description: ${t.description}\n` +
      `  - sender hints: ${senders}\n` +
      `  - keyword hints: ${keywords}\n` +
      `  - fields to extract: [${fields}]`
    );
  }).join("\n\n");
}

File: agentmail-approval-inbox/python/types_config.py
"""
Loads approval_types.yaml and exposes the type list to the classifier prompt
and the action runner. Re-loads on every call so live edits to the YAML take
effect without restarting the agent.
"""

from dataclasses import dataclass
from pathlib import Path

import yaml

CONFIG_FILE = Path("approval_types.yaml")


@dataclass
class TypeConfig:
    type: str
    description: str
    classifier_hints: dict  # {senders: [], keywords: []}
    extract_fields: list[str]
    approve: dict  # {forward_to?, webhook?, reply_to_sender?}
    decline: dict


def load() -> list[TypeConfig]:
    if not CONFIG_FILE.exists():
        return []
    try:
        data = yaml.safe_load(CONFIG_FILE.read_text())
    except Exception as e:
        print(f"  ! approval_types.yaml malformed: {e}")
        return []
    out: list[TypeConfig] = []
    for entry in (data or {}).get("types", []) or []:
        out.append(TypeConfig(
            type=entry["type"],
            description=entry.get("description", ""),
            classifier_hints=entry.get("classifier_hints", {}) or {},
            extract_fields=entry.get("extract_fields", []) or [],
            approve=entry.get("approve", {}) or {},
            decline=entry.get("decline", {}) or {},
        ))
    return out


def find(types: list[TypeConfig], type_name: str) -> TypeConfig | None:
    for t in types:
        if t.type == type_name:
            return t
    return None


def render_for_prompt(types: list[TypeConfig]) -> str:
    """Render the configured types as a readable block for Claude's classifier prompt."""
    if not types:
        return "(no request types configured — every email will be discarded)"
    lines = []
    for t in types:
        senders = ", ".join(t.classifier_hints.get("senders", []) or []) or "(no sender hints)"
        keywords = ", ".join(t.classifier_hints.get("keywords", []) or []) or "(no keyword hints)"
        fields = ", ".join(t.extract_fields) or "(no fields)"
        lines.append(
            f"## {t.type}\n"
            f"  - description: {t.description}\n"
            f"  - sender hints: {senders}\n"
            f"  - keyword hints: {keywords}\n"
            f"  - fields to extract: [{fields}]"
        )
    return "\n\n".join(lines)
